### PR TITLE
docs: addition of param and result sections

### PR DESCRIPTION
**Pull request by hachreak**
* Improves documentation. (addresses #95)

Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

File: invenio_deposit/api.py [hunk 1]
```diff
@@ -76,7 +80,11 @@ def wrapper(self_or_cls, *args, **kwargs):
 
 
 def has_status(method=None, status='draft'):
-    """Check that deposit has defined status (default: draft)."""
+    """Check that deposit has a defined status (default: draft).
+
+    :param method: Function executed if record has a defined status.
+    :params status: Defined status to check.
```
Comment (greut) [inline invenio_deposit/api.py]: I'd use `param` instead of `params`.

http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists

Comment (hachreak) [inline invenio_deposit/api.py]: ups.. you are right... I'll change all XD

**Comment by hachreak** [comment]
```python
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/sphinx/cmdline.py", line 244, in main
    app.build(opts.force_all, filenames)
  File "/usr/local/lib/python2.7/site-packages/sphinx/application.py", line 297, in build
    self.builder.build_update()
  File "/usr/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 251, in build_update
    'out of date' % len(to_build))
  File "/usr/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 322, in build
    self.write(docnames, list(updated_docnames), method)
  File "/usr/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 360, in write
    self._write_serial(sorted(docnames), warnings)
  File "/usr/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 370, in _write_serial
    self.warn(*warning, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sphinx/application.py", line 360, in warn
    raise SphinxWarning(warntext)
SphinxWarning: :1: WARNING: py:class reference target not found: Record
```
@jirikuncar 
any clue where is the problem? :)

**Comment by jirikuncar** [comment]
@hachreak you need to configure intersphinx in `docs/conf.py`.

File: docs/conf.py [hunk 2]
```diff
@@ -328,4 +328,8 @@
 
 
 # Example configuration for intersphinx: refer to the Python standard library.
-intersphinx_mapping = {'https://docs.python.org/': None}
+intersphinx_mapping = {
+    'https://docs.python.org/': None,
+    'invenio-pidstore': ('http://pythonhosted.org/invenio-pidstore', None),
+    'invenio-records': ('http://pythonhosted.org/invenio-records', None),
```
Comment (jirikuncar) [inline docs/conf.py]: You probably need also `invenio-records-files`.

File: docs/api.rst [hunk 3]
```diff
@@ -28,3 +28,5 @@ API Docs
 invenio_deposit
```
Comment (jirikuncar) [inline docs/api.rst]: IMHO we can remove lines 28-30.

File: docs/conf.py (hunk 2 above)
Comment (hachreak) [inline docs/conf.py]: done

**Comment by jirikuncar** [comment]
@hachreak what about `docs: addition of param and result sections`?

**Comment by hachreak** [comment]
> any clue where is the problem? :)
>> @hachreak you need to configure intersphinx in docs/conf.py. (as you have probably done)

I configured also `invenio-records-files`, but still not working..

File: invenio_deposit/api.py [hunk 4]
```diff
@@ -56,7 +56,11 @@
 
 
 def index(method=None, delete=False):
-    """Update index."""
+    """Decorator to update index.
+
+    :param method: Function wrapped. (Default: None).
```
Comment (jirikuncar) [inline invenio_deposit/api.py]:
```
. (Default: ``None``)
. (Default: ``'draft'``)
```

1. After dot then I would not add another one after brackets.
2. Inside param description (before last dot).

File: invenio_deposit/api.py [hunk 5]
```diff
@@ -258,10 +298,39 @@ def _publish_edited(self):
...
+
+        * A dump of all information inside the deposit is done.
+
+        * A snapshot of the files is done.
+
+        Otherwise, published the new edited version.
+        In this case, if in the mainwhile someone already published a new
+        version, it'll try to merge the changes with the latest version.
+
+        Note: no need for indexing as it calls `self.commit()`.
```
Comment (jirikuncar) [inline invenio_deposit/api.py]:
```
.. note::
```

File: invenio_deposit/api.py [hunk 6]
```diff
@@ -289,7 +361,32 @@ def _prepare_edit(self, record):
     @has_status(status='published')
     @index
     def edit(self, pid=None):
-        """Edit deposit."""
+        """Edit deposit.
+
+        #. The signal :data:`invenio_records.signals.before_record_update`
+            is sent before the edit execution.
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: Can you check the indentation here?

File: invenio_deposit/api.py [hunk 7]
```diff
@@ -310,7 +407,30 @@ def edit(self, pid=None):
...
+
+            deposit['_deposit']['status'] = 'draft'
+            deposit['$schema'] = deposit_schema_from_record_schema
+
+        #. The signal :data:`invenio_records.signals.after_record_update` is
+            sent after the edit execution.
+
+        #. The deposit index is updated.
+
+        Status required: `draft`.
```
Comment (jirikuncar) [inline invenio_deposit/api.py]:
```
'draft'
# or better
``'draft'``
```

File: invenio_deposit/api.py [hunk 8]
```diff
@@ -365,7 +507,10 @@ def _create_bucket(self):
 
     @property
     def files(self):
-        """Add validation on ``sort_by`` method."""
+        """List of Files inside the deposit.
+
+        Add validation on ``sort_by`` method.
```
Comment (jirikuncar) [inline invenio_deposit/api.py]: Can you elaborate on this sentence?

File: invenio_deposit/config.py [hunk 9]
```diff
@@ -100,13 +106,24 @@
...
 
 DEPOSIT_REST_DEFAULT_SORT = dict(
     deposits=dict(
         query='bestmatch',
         noquery='mostrecent'
     )
 )
+"""Default deposit sort configuration.
+
+See class:`invenio_records_rest.config:RECORDS_REST_DEFAULT_SORT` for more
```
Comment (jirikuncar) [inline invenio_deposit/config.py]: `:class:`

File: invenio_deposit/config.py [hunk 10]
```diff
@@ -120,6 +137,12 @@
         )
     )
 )
+"""Basic deposit facts configuration.
+
+See class:`invenio_records_rest.config:RECORDS_REST_FACETS` for more
```
Comment (jirikuncar) [inline invenio_deposit/config.py]: `:class:`?

File: invenio_deposit/ext.py [hunk 11]
```diff
@@ -42,7 +42,13 @@ def __init__(self, app=None):
             self.init_app(app)
 
     def init_app(self, app):
-        """Flask application initialization."""
+        """Flask application initialization.
+
+        Initialize the CLI and all UI endpoints.
+        Connect all signals if `DEPOSIT_REGISTER_SIGNALS` is True.
```
Comment (jirikuncar) [inline invenio_deposit/ext.py]: double \`

File: invenio_deposit/ext.py [hunk 12]
```diff
@@ -42,7 +42,13 @@ def __init__(self, app=None):
...
+
+        :param app: The Flask application.
```
Comment (jirikuncar) [inline invenio_deposit/ext.py]: What about?
```
An instance of :class:`~flask.app.Flask`.
```

File: docs/api.rst [hunk 13]
```diff
@@ -25,6 +25,5 @@
 API Docs
 ========
 
-invenio_deposit
----------------
-
+.. automodule:: invenio_deposit.api
```
Comment (jirikuncar) [inline docs/api.rst]: can you include also other files like views and providers?

Comment (hachreak) [inline docs/api.rst]: How should I strucuture the docs?

see https://github.com/inveniosoftware/invenio-deposit/issues/96

Comment (jirikuncar) [inline docs/api.rst]: For now just add `.. automodule:: invenio.{views, providers, etc}`. You can include also a titles:

```
Views
-----
```

Comment (hachreak) [inline docs/api.rst]: @jirikuncar 
Sphinx said me: 
```
WARNING: py:data reference target not found: invenio_records_rest.config.RECORDS_REST_DEFAULT_SORT
```

But, actually, this configuration is not documented online for now (see my https://github.com/inveniosoftware/invenio-records-rest/pull/105 ).
Is it why sphinx returns the warning?

Comment (jirikuncar) [inline docs/api.rst]: Yes. We need to document it in invenio-records-rest and upload new docs on pythonhosted.

Comment (hachreak) [inline docs/api.rst]: So, for my PR for `invenio-records-rest` I need before at least `invenio_search.api.RecordsSearch` is available online.

Comment (jirikuncar) [inline docs/api.rst]: @hachreak will you fix `invenio_search` first?

**Comment by hachreak** [comment]
blocked by: https://github.com/inveniosoftware/invenio-records-rest/pull/105

File: invenio_deposit/api.py [hunk 14]
```diff
@@ -91,7 +100,13 @@ def wrapper(self, *args, **kwargs):
 
 
 def preserve(method=None, result=True, fields=None):
-    """Preserve fields in deposit."""
+    """Preserve fields in deposit.
+
+    :param method: Function to execute. (Default: ``None``)
+    :param result: If `True` returns the result of method execution,
+        otherwise `self`. (Default: ``True``)
+    :param fields: List of fields to preserve (default: ('_deposit',)).
```
Comment (jirikuncar) [inline invenio_deposit/api.py]:
```
(Default: ``('_deposit', )``).
```

File: invenio_deposit/api.py [hunk 15]
```diff
@@ -76,7 +80,12 @@ def wrapper(self_or_cls, *args, **kwargs):
 
 
 def has_status(method=None, status='draft'):
-    """Check that deposit has defined status (default: draft)."""
+    """Check that deposit has a defined status (default: draft).
+
+    :param method: Function executed if record has a defined status.
+        (Default: ``None``)
+    :param status: Defined status to check. (Default: 'draft')
```
Comment (jirikuncar) [inline invenio_deposit/api.py]:
```
``'draft'``
```

File: invenio_deposit/links.py [hunk 16]
```diff
@@ -29,7 +29,23 @@
...
 def deposit_links_factory(pid):
-    """Factory for record links generation."""
+    """Factory for record links generation.
+
+    The dictionary is formed as:
+
+    .. code-block:: python
+
+        {
+            "files": "/url/to/files",
```
Comment (jirikuncar) [inline invenio_deposit/links.py]: `s/"/'/`

**Comment by hachreak** [comment]
ping @jirikuncar

**Pull request merged**